src/main/java/com/conveyal/datatools/manager/models/transform/PreserveCustomFieldsTransformation.java

```diff
@@ -0,0 +1,134 @@
+        String tableName = table + ".txt";
+        Path targetZipPath = Paths.get(zipTarget.gtfsFile.getAbsolutePath());
+
+        Table specTable = Arrays.stream(Table.tablesInOrder)
+            .filter(t -> t.name.equals(table))
+            .findFirst()
+            .get();
+        List<String> specTableFields = specTable.specFields().stream().map(f -> f.name).collect(Collectors.toList());
+        List<String> tablePrimaryKeys = specTable.getPrimaryKeyNames();
+
+        try (FileSystem targetZipFs = FileSystems.newFileSystem( targetZipPath, (ClassLoader) null )) {
+            Path targetTxtFilePath = getTablePathInZip(tableName, targetZipFs);
+
+            final File tempFile = File.createTempFile(tableName + "-temp", ".txt");
+            File output = File.createTempFile(tableName + "-output-temp", ".txt");
+            int rowsModified = 0;
+            List<String> customFields;
+
+            try (
+                InputStream is = Files.newInputStream(targetTxtFilePath);
+                CsvMapReader customFileReader = new CsvMapReader(new StringReader(csvData), CsvPreference.STANDARD_PREFERENCE);
+                CsvMapReader editorFileReader = new CsvMapReader(new InputStreamReader(is), CsvPreference.STANDARD_PREFERENCE);
+                CsvMapWriter writer = new CsvMapWriter(new FileWriter(output), CsvPreference.STANDARD_PREFERENCE);
+            ){
+
+                String[] customHeaders = customFileReader.getHeader(true);
+                String[] editorHeaders = editorFileReader.getHeader(true);
+
+                customFields = Arrays.stream(customHeaders).filter(h -> !specTableFields.contains(h)).collect(Collectors.toList());
+                if (customFields.isEmpty()) return;
+                String[] fullHeaders = ArrayUtils.addAll(editorHeaders, customFields.toArray(new String[0]));
+
+                HashMap<String, Map<String, String>> customFieldsLookup = createCsvHashMap(customFileReader, customHeaders, tablePrimaryKeys);
+                writer.writeHeader(fullHeaders);
+
+                Map<String, String> row;
+                while ((row = editorFileReader.read(editorHeaders)) != null) {
+                    List<String> editorCsvPrimaryKeyValues = tablePrimaryKeys.stream()
+                        .map(row::get)
+                        .collect(Collectors.toList());
+
+                    String hashKey = StringUtils.join(editorCsvPrimaryKeyValues, "_");
+                    Map<String, String> customCsvValues = customFieldsLookup.get(hashKey);
+                    Map<String, String> finalRow = row;
+                    customFields.stream().forEach(customField -> {
+                        String value = customCsvValues == null ? null : customCsvValues.get(customField);
+                        finalRow.put(customField, value);
+                    });
+                    if (customCsvValues != null) rowsModified++;
```

Comment:
This if condition is separate right now because we want to write null values in the row if `customCsvValues` is null. I can rework it to write nulls for each customField if `customCsvValues` is null but I'm not sure that simplifies things that much.

Reply:
Is it because you need to have an entry in the map in order to print a bank column if the custom field is not set?

Reply:
Yep exactly